src/ActionList/Group.tsx

```diff
@@ -58,32 +79,64 @@ export const Group: React.FC<React.PropsWithChildren<ActionListGroupProps>> = ({
       }}
       {...props}
     >
-      {title && <Header title={title} variant={variant} auxiliaryText={auxiliaryText} labelId={labelId} />}
+      {(slots.groupHeading || title) && (
+        <GroupHeading
+          title={title}
+          variant={variant}
+          auxiliaryText={auxiliaryText}
+          groupHeadingId={groupHeadingId}
+          as={slots.groupHeading?.props.as}
+        >
+          {slots.groupHeading ? slots.groupHeading.props.children : null}
+        </GroupHeading>
+      )}
       <GroupContext.Provider value={{selectionVariant}}>
         <Box
           as="ul"
           sx={{paddingInlineStart: 0}}
-          aria-labelledby={title ? labelId : undefined}
+          // if listRole is set (listbox or menu), we don't label the list with the groupHeadingId
+          // because the heading is hidden from the accessibility tree and only used for presentation role.
+          // We will instead use aria-label to label the list. See a line below.
+          aria-labelledby={listRole ? undefined : groupHeadingId}
+          aria-label={listRole ? title ?? (slots.groupHeading?.props.children as string) : undefined}
           role={role || (listRole && 'group')}
         >
-          {props.children}
+          {slots.groupHeading ? childrenWithoutSlots : props.children}
         </Box>
       </GroupContext.Provider>
     </Box>
   )
 }
 
-export type HeaderProps = Pick<ActionListGroupProps, 'variant' | 'title' | 'auxiliaryText'> & {
-  labelId: string
-}
+export type GroupHeadingProps = Pick<ActionListGroupProps, 'variant' | 'title' | 'auxiliaryText'> &
+  Omit<ActionListHeadingProps, 'as'> &
+  SxProp &
+  React.HTMLAttributes<HTMLElement> & {
+    as?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
+    groupHeadingId?: string
```

Comment:
Is it possible for someone to pass this as a prop and have it update properly in the parent component? If not, or we don't want folks to be passing it, would this makes sense as something that is read from context?

Reply:
Ooo great point 🔥 Yes we don't want folks to be passing this id in, so totally makes sense to get it from context!

Reply:
committed [here](https://github.com/primer/react/pull/3900/commits/b929c0046bffc86cc94625993afc2c398cf49516) 🙂